File: lib/models/message.dart
enum MessageType {
  text,
  image,
  document,
  audio,
  video,
}

enum MessageStatus {
  sending,
  sent,
  delivered,
  read,
  failed,
}

class Message {
  final String id;
  final String chatId;
  final String senderId;
  final String? receiverId; // Null for group messages
  final String? groupId; // Null for one-to-one messages
  final MessageType type;
  final String? encryptedContent;
  final String? mediaUrl;
  final MessageStatus status;
  final DateTime timestamp;
  final DateTime? expiresAt; // For self-destructing messages
  final bool isDeleted;
  final bool isAnonymous; // Used for anonymous messages in groups
  final Map<String, dynamic>? metadata; // Additional info for attachments

  Message({
    required this.id,
    required this.chatId,
    required this.senderId,
    this.receiverId,
    this.groupId,
    required this.type,
    this.encryptedContent,
    this.mediaUrl,
    required this.status,
    required this.timestamp,
    this.expiresAt,
    this.isDeleted = false,
    this.isAnonymous = false,
    this.metadata,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['_id'] ?? json['id'],
      chatId: json['chatId'],
      senderId: json['senderId'],
      receiverId: json['receiverId'],
      groupId: json['groupId'],
      type: MessageType.values.firstWhere(
        (e) => e.toString() == 'MessageType.${json['type']}',
        orElse: () => MessageType.text,
      ),
      encryptedContent: json['encryptedContent'],
      mediaUrl: json['mediaUrl'],
      status: MessageStatus.values.firstWhere(
        (e) => e.toString() == 'MessageStatus.${json['status']}',
        orElse: () => MessageStatus.sent,
      ),
      timestamp: DateTime.parse(json['timestamp']),
      expiresAt: json['expiresAt'] != null
          ? DateTime.parse(json['expiresAt'])
          : null,
      isDeleted: json['isDeleted'] ?? false,
      isAnonymous: json['isAnonymous'] ?? false,
      metadata: json['metadata'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'chatId': chatId,
      'senderId': senderId,
      'receiverId': receiverId,
      'groupId': groupId,
      'type': type.toString().split('.').last,
      'encryptedContent': encryptedContent,
      'mediaUrl': mediaUrl,
      'status': status.toString().split('.').last,
      'timestamp': timestamp.toIso8601String(),
      'expiresAt': expiresAt?.toIso8601String(),
      'isDeleted': isDeleted,
      'isAnonymous': isAnonymous,
      'metadata': metadata,
    };
  }

  bool get isExpired {
    if (expiresAt == null) return false;
    return DateTime.now().isAfter(expiresAt!);
  }

  bool get isMedia {
    return type != MessageType.text;
  }

  Message copyWith({
    String? id,
    String? chatId,
    String? senderId,
    String? receiverId,
    String? groupId,
    MessageType? type,
    String? encryptedContent,
    String? mediaUrl,
    MessageStatus? status,
    DateTime? timestamp,
    DateTime? expiresAt,
    bool? isDeleted,
    bool? isAnonymous,
    Map<String, dynamic>? metadata,
  }) {
    return Message(
      id: id ?? this.id,
      chatId: chatId ?? this.chatId,
      senderId: senderId ?? this.senderId,
      receiverId: receiverId ?? this.receiverId,
      groupId: groupId ?? this.groupId,
      type: type ?? this.type,
      encryptedContent: encryptedContent ?? this.encryptedContent,
      mediaUrl: mediaUrl ?? this.mediaUrl,
      status: status ?? this.status,
      timestamp: timestamp ?? this.timestamp,
      expiresAt: expiresAt ?? this.expiresAt,
      isDeleted: isDeleted ?? this.isDeleted,
      isAnonymous: isAnonymous ?? this.isAnonymous,
      metadata: metadata ?? this.metadata,
    );
  }
}

File: server/models/message.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const messageSchema = new Schema({
  chatId: {
    type: Schema.Types.ObjectId,
    ref: 'Chat',
    required: true,
  },
  senderId: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  receiverId: {
    type: Schema.Types.ObjectId,
    ref: 'User',
  },
  groupId: {
    type: Schema.Types.ObjectId,
    ref: 'Group',
  },
  type: {
    type: String,
    enum: ['text', 'image', 'document', 'audio', 'video'],
    default: 'text',
  },
  encryptedContent: {
    type: String,
  },
  mediaUrl: {
    type: String,
  },
  status: {
    type: String,
    enum: ['sending', 'sent', 'delivered', 'read', 'failed'],
    default: 'sent',
  },
  timestamp: {
    type: Date,
    default: Date.now,
  },
  expiresAt: {
    type: Date,
  },
  isDeleted: {
    type: Boolean,
    default: false,
  },
  isAnonymous: {
    type: Boolean,
    default: false,
  },
  metadata: {
    type: Object,
    default: {},
  },
  readBy: [{
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
  }],
  deliveredTo: [{
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
  }],
}, {
  timestamps: true
});

// Virtual for getting ID as a string
messageSchema.virtual('id').get(function() {
  return this._id.toHexString();
});

// Ensure virtual fields are serialized
messageSchema.set('toJSON', {
  virtuals: true,
  transform: (doc, ret) => {
    delete ret._id;
    delete ret.__v;
    return ret;
  }
});

// Index to help query expiring messages
messageSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });

// Method to check if a message is expired
messageSchema.methods.isExpired = function() {
  if (!this.expiresAt) {
    return false;
  }
  return new Date() > this.expiresAt;
};

// Method to mark message as read by a user
messageSchema.methods.markAsReadBy = function(userId) {
  if (!this.readBy.some(read => read.userId.toString() === userId.toString())) {
    this.readBy.push({ userId, timestamp: new Date() });
    this.status = 'read';
    return this.save();
  }
  return Promise.resolve(this);
};

// Method to mark message as delivered to a user
messageSchema.methods.markAsDeliveredTo = function(userId) {
  if (!this.deliveredTo.some(delivered => delivered.userId.toString() === userId.toString())) {
    this.deliveredTo.push({ userId, timestamp: new Date() });
    if (this.status === 'sent') {
      this.status = 'delivered';
    }
    return this.save();
  }
  return Promise.resolve(this);
};

module.exports = mongoose.model('Message', messageSchema);